chain/substreams/src/mapper.rs

```diff
@@ -89,3 +113,237 @@ impl SubstreamsMapper<Chain> for Mapper {
         }
     }
 }
+
+fn parse_changes(
+    changes: &EntityChanges,
+    schema: &Arc<InputSchema>,
+) -> anyhow::Result<Vec<ParsedChanges>> {
+    let mut parsed_changes = vec![];
+    for entity_change in changes.entity_changes.iter() {
+        let mut parsed_data: HashMap<Word, Value> = HashMap::default();
+        let entity_type = EntityType::new(entity_change.entity.to_string());
+
+        // Make sure that the `entity_id` gets set to a value
+        // that is safe for roundtrips through the database. In
+        // particular, if the type of the id is `Bytes`, we have
+        // to make sure that the `entity_id` starts with `0x` as
+        // that will be what the key for such an entity have
+        // when it is read from the database.
+        //
+        // Needless to say, this is a very ugly hack, and the
+        // real fix is what's described in [this
+        // issue](https://github.com/graphprotocol/graph-node/issues/4663)
+        let entity_id: String = match schema.id_type(&entity_type)? {
+            IdType::String => entity_change.id.clone(),
+            IdType::Bytes => {
+                if entity_change.id.starts_with("0x") {
+                    entity_change.id.clone()
+                } else {
+                    format!("0x{}", entity_change.id)
+                }
+            }
+        };
+        let key = EntityKey {
+            entity_type,
+            entity_id: Word::from(entity_id),
+            causality_region: CausalityRegion::ONCHAIN, // Substreams don't currently support offchain data
+        };
+        let id = schema.id_value(&key)?;
+        parsed_data.insert(Word::from("id"), id);
+
+        let changes = match entity_change.operation() {
+            entity_change::Operation::Create | entity_change::Operation::Update => {
+                for field in entity_change.fields.iter() {
+                    let new_value: &codec::value::Typed = match &field.new_value {
+                        Some(codec::Value {
+                            typed: Some(new_value),
+                        }) => &new_value,
+                        _ => continue,
+                    };
+
+                    let value: Value = decode_value(new_value)?;
+                    *parsed_data
+                        .entry(Word::from(field.name.as_str()))
+                        .or_insert(Value::Null) = value;
+                }
+                let entity = schema
+                    .make_entity(parsed_data)
+                    .map_err(anyhow::Error::from)?;
+
+                ParsedChanges::Upsert { key, entity }
+            }
+            entity_change::Operation::Delete => ParsedChanges::Delete(key),
+            entity_change::Operation::Unset => ParsedChanges::Unset,
```

Comment:
I am not familiar with substreams vocab, but what does `Unset` mean in this context? Do we even need to keep that in the parsed changes? Might be good to add that as a comment to `ParsedChanges`

Reply:
will do, Unset is just to preserve the code logic that's applied to protobuf. Unset there is the default value when a new unsupported value is used. Say you have an enum with 3 values,  if the server has a 4th value then when parsing it will default to Unset, works as a sentinel value of sorts. 

In this case, at parsing we don't "care" about what operation it is, we're just transforming so there is a 1:1 mapping of supported operations. Once we run the trigger processing then there will be an error if the operation is `Unset` as that's not a valid operation. I didn't want to change the responsibility of the block stream so this was the "easiest" way to maintain it

Reply:
I will also add a comment here to make it clearer